### PR TITLE
Fix 275. Fix edgecase where it was not possible to patch only termini of a specific chain

### LINKED ISSUE #275: AnnoteMutMod fails with water present in pdb
When water or any 1 residue molecule is present in the pdb file annotation of modifications and mutations fails, due to the fact that there are no neighbors in a residue graph with only one residue. Specifically the following line fails:

https://github.com/marrink-lab/vermouth-martinize/blob/67cfeb324a1b9c7c6d385f080766072a9802d7df/vermouth/processors/annotate_mut_mod.py#L111

### PR DESCRIPTION
**Pull request by pckroon**
In addition, split of the termini matching logic into a separate function.

Fixes #275

### CODE REVIEW

File: vermouth/processors/annotate_mut_mod.py [hunk 1]
```diff
@@ -108,15 +108,52 @@ def residue_matches(resspec, residue_graph, res_idx):
...
+        resspec.pop('resname', '')
+        resspec.pop('resid', 0)
+    return _subdict(resspec, residue)
+
+
+def _terminal_matches(resname, residue_graph, res_idx):
+    """
+    Tests whether the node res_idx in graph residue_graph is a cter or nter, as
+    determined by resname. Resname must be one of nter or cter.
+    It is assumed that the degree of the specified node is 1.
```
Comment (fgrunewald) [inline vermouth/processors/annotate_mut_mod.py:126]: Since this is already a seperate function you could also check and/or raise an error?!

File: vermouth/processors/annotate_mut_mod.py [hunk 2]
```diff
@@ -108,15 +108,52 @@ def residue_matches(resspec, residue_graph, res_idx):
...
     #        more clever, addressing arbitrarily branched polymers and
     #        termini
+    if residue_graph.degree[res_idx] == 1 and resspec.get('resname') in ['nter', 'cter']:
+        if not _terminal_matches(resspec.get('resname'), residue_graph, res_idx):
+            return False
+        # Remove resname and resid from the resspec, since they have special meaning
+        # in this case.
+        resspec = resspec.copy()
+        resspec.pop('resname', '')
+        resspec.pop('resid', 0)
```
Comment (fgrunewald) [inline vermouth/processors/annotate_mut_mod.py:118]: Why this addition? It wasn't there in the previous code, was it.

File: vermouth/tests/test_annotate_mut_mod.py [hunk 3]
```diff
@@ -142,6 +153,23 @@ def test_annotate_modifications(example_mol, modifications, mutations, expected_
...
+    mol.add_nodes_from(enumerate(nodes))
+    mol.add_edges_from([(0, 1)])
+
+    modification = [({'resname': 'A', 'resid': 2}, 'C-ter'),]
+    annotate_modifications(mol, modification, [])
+
+    assert mol.nodes[0] == {'modification': ['C-ter'], 'resname': 'A', 'resid': 2, 'chain': 'A'}
+    assert mol.nodes[1] == {'modification': ['C-ter'], 'resname': 'A', 'resid': 2, 'chain': 'A'}
+
+
```
Comment (fgrunewald) [inline vermouth/tests/test_annotate_mut_mod.py:172]: I'm of the opinion this should raise an error because or a single amino acids C and N terminus are ill-defined terms in the martini sense, aren't they.

**Comment by pckroon** [comment]
#276 is out of scope for this PR, and is fixed in #278

**Comment by pckroon** [comment]
This is still relevant, and tests are included. Or I'm confused on what edge cases you mean.

**Comment by fgrunewald** [comment]
Hey, I mean you called the PR "Fix edgecase". But I only saw a test on the single molecule I believe. So this PR was solely for that purpose? And then it got closed and reopened so I think I am as confused as you are.

**Comment by pckroon** [comment]
The title of the PR is incorrect/outdated. This PR deals with applying terminal mutations to single residue molecules.
